sql: TestDropTableDeleteData failed

sql.TestDropTableDeleteData [failed](https://tanzanite.cluster.engflow.com/invocations/default/2615b65c-c9e5-40ae-a3f3-81a811bc2432?testReportRun=1&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [546218e68c128176c7108fd2fe1dd213e41db60b](https://github.com/cockroachdb/cockroach/commits/546218e68c128176c7108fd2fe1dd213e41db60b):

Fatal error:

```
panic: test timed out after 1h16m37s
running tests:
	TestDropTableDeleteData (1h15m44s)

```
Stack: 

```
goroutine 369584 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue.func2()
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:425 +0xbd
created by github.com/cockroachdb/cockroach/pkg/util/admission.initWorkQueue in goroutine 23680
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:422 +0x676

goroutine 23975 [select]:
github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:502 +0x1cd
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 23680
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:473 +0x8e

goroutine 23781 [select]:
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:660 +0x19b
created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked in goroutine 23680
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:649 +0x105

goroutine 23779 [select, 75 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc0048eaa80, {0xdf90970, 0xc0065b0370})
	github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0xd3
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig in goroutine 23680
	github.com/cockroachdb/cockroach/pkg/util/log/flags.go:319 +0x1d65

goroutine 23810 [select]:
github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs.func1()
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:502 +0x1cd
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*WorkQueue).startClosingEpochs in goroutine 23680
	github.com/cockroachdb/cockroach/pkg/util/admission/work_queue.go:473 +0x8e

goroutine 23812 [select]:
github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:660 +0x19b
created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked in goroutine 23680
	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:649 +0x105

goroutine 261926 [sync.Cond.Wait, 74 minutes]:
sync.runtime_notifyListWait(0xc008c220a8, 0x5b6)
	GOROOT/src/runtime/sema.go:569 +0x159
sync.(*Cond).Wait(0xc008c22098)
	GOROOT/src/sync/cond.go:70 +0x75
github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:369
github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0xc008c22000, {0xc00a245510?, 0x977a3e0?})
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:643 +0xb79
runtime/pprof.Do({0xdf8d5f8, 0x13718040}, {{0xc001201000?, 0x20808e0?, 0xc00197b7e0?}}, 0xc005f36fc0)
	GOROOT/src/runtime/pprof/runtime.go:51 +0x118
github.com/cockroachdb/pebble/record.NewLogWriter.func2()
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:562 +0x92
created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 23855
	github.com/cockroachdb/pebble/record/external/com_github_cockroachdb_pebble/record/log_writer.go:561 +0x8e5
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDropTableDeleteData.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36714

--- a/pkg/sql/drop_test.go
+++ b/pkg/sql/drop_test.go
@@ -693,8 +693,9 @@ func TestDropTableDeleteData(t *testing.T) {
 		},
 	}
 
+	ctx := context.Background()
 	srv, sqlDB, kvDB := serverutils.StartServer(t, params)
-	defer srv.Stopper().Stop(context.Background())
+	defer srv.Stopper().Stop(ctx)
 	s := srv.ApplicationLayer()
 	codec := s.Codec()
 	systemDB := srv.SystemLayer().SQLConn(t)
@@ -710,57 +711,59 @@ func TestDropTableDeleteData(t *testing.T) {
 	const numKeys = 3 * numRows
 	const numTables = 5
 	var descs []catalog.TableDescriptor
-	for i := 0; i < numTables; i++ {
-		tableName := fmt.Sprintf("test%d", i)
-		if err := tests.CreateKVTable(sqlDB, tableName, numRows); err != nil {
-			t.Fatal(err)
+
+	func() {
+		defer close(allowGC)
+		for i := 0; i < numTables; i++ {
+			tableName := fmt.Sprintf("test%d", i)
+			if err := tests.CreateKVTable(sqlDB, tableName, numRows); err != nil {
+				t.Fatal(err)
+			}
+
+			descs = append(descs, desctestutils.TestingGetPublicTableDescriptor(kvDB, codec, "t", tableName))
+
+			parentDatabaseID := descpb.ID(sqlutils.QueryDatabaseID(t, sqlDB, "t"))
+			parentSchemaID := descpb.ID(sqlutils.QuerySchemaID(t, sqlDB, "t", "public"))
+
+			nameKey := catalogkeys.EncodeNameKey(codec, &descpb.NameInfo{
+				ParentID:       parentDatabaseID,
+				ParentSchemaID: parentSchemaID,
+				Name:           tableName,
+			})
+			gr, err := kvDB.Get(ctx, nameKey)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if !gr.Exists() {
+				t.Fatalf("Name entry %q does not exist", nameKey)
+			}
+
+			tableSpan := descs[i].TableSpan(codec)
+			tests.CheckKeyCount(t, kvDB, tableSpan, numKeys)
+
+			if _, err := sqltestutils.AddImmediateGCZoneConfig(sqlDB, descs[i].GetID()); err != nil {
+				t.Fatal(err)
+			}
 		}
 
-		descs = append(descs, desctestutils.TestingGetPublicTableDescriptor(kvDB, codec, "t", tableName))
-
-		parentDatabaseID := descpb.ID(sqlutils.QueryDatabaseID(t, sqlDB, "t"))
-		parentSchemaID := descpb.ID(sqlutils.QuerySchemaID(t, sqlDB, "t", "public"))
-
-		nameKey := catalogkeys.EncodeNameKey(codec, &descpb.NameInfo{
-			ParentID:       parentDatabaseID,
-			ParentSchemaID: parentSchemaID,
-			Name:           tableName,
-		})
-		gr, err := kvDB.Get(ctx, nameKey)
-		if err != nil {
-			t.Fatal(err)
-		}
-		if !gr.Exists() {
-			t.Fatalf("Name entry %q does not exist", nameKey)
+		for i := 0; i < numTables; i++ {
+			if _, err := sqlDB.Exec(fmt.Sprintf(`DROP TABLE t.%s`, descs[i].GetName())); err != nil {
+				t.Fatal(err)
+			}
 		}
 
-		tableSpan := descs[i].TableSpan(codec)
-		tests.CheckKeyCount(t, kvDB, tableSpan, numKeys)
-
-		if _, err := sqltestutils.AddImmediateGCZoneConfig(sqlDB, descs[i].GetID()); err != nil {
-			t.Fatal(err)
+		// Data hasn't been GC-ed.
+		for i := 0; i < numTables; i++ {
+			if err := descExists(sqlDB, true, descs[i].GetID()); err != nil {
+				t.Fatalf("table (id=%d name=%s) was deleted. error: %v", descs[i].GetID(), descs[i].GetName(), err)
+			}
+			tableSpan := descs[i].TableSpan(codec)
+			tests.CheckKeyCountIncludingTombstoned(t, srv.StorageLayer(), tableSpan, numKeys)
 		}
-	}
-
-	for i := 0; i < numTables; i++ {
-		if _, err := sqlDB.Exec(fmt.Sprintf(`DROP TABLE t.%s`, descs[i].GetName())); err != nil {
-			t.Fatal(err)
-		}
-	}
-
-	// Data hasn't been GC-ed.
-	sqlRun := sqlutils.MakeSQLRunner(sqlDB)
-	for i := 0; i < numTables; i++ {
-		if err := descExists(sqlDB, true, descs[i].GetID()); err != nil {
-			t.Fatalf("table (id=%d name=%s) was deleted. error: %v", descs[i].GetID(), descs[i].GetName(), err)
-		}
-		tableSpan := descs[i].TableSpan(codec)
-		tests.CheckKeyCountIncludingTombstoned(t, srv.StorageLayer(), tableSpan, numKeys)
-	}
-
-	close(allowGC)
+	}()
 
 	// Now verify that the GC job has run.
+	sqlRun := sqlutils.MakeSQLRunner(sqlDB)
 	for i := 0; i < numTables; i++ {
 		if err := jobutils.VerifySystemJob(t, sqlRun, numTables+i,
 			jobspb.TypeNewSchemaChange, jobs.StatusSucceeded, jobs.Record{

